File: lib_js/models/supabase/sources.js
import { SupabaseModel, Enum } from "./supabaseModel.js";

// Define SourceTypeEnum
class SourceTypeEnum extends Enum {
    constructor() {
        super();
        this.DOCUMENT = "document";
        this.WEBPAGE = "webpage";
        this.WEBSITE = "website";
        this.VIDEO = "video";
        this.AUDIO = "audio";
        this.IMAGE = "image";
        this.TOPIC = "topic";
        this.CONCEPT = "concept";
        this.COLLECTION = "collection";
        Object.freeze(this);
    }
}

export const SourceType = new SourceTypeEnum();

// Define SourceModel
export class SourceModel extends SupabaseModel {
    static TABLE_NAME = "source";
    static TABLE_FIELDS = {
        id: { type: "uuid", required: true, dbColumn: "id" },
        type: { type: SourceType, required: false, dbColumn: "type" },
        originalSource: {
            type: "string",
            required: false,
            dbColumn: "original_source"
        },
        resolvedSource: {
            type: "string",
            required: false,
            dbColumn: "resolved_source"
        },
        title: { type: "string", required: false, dbColumn: "title" },
        lang: { type: "string", required: false, dbColumn: "lang" },
        contentHash: {
            type: "string",
            required: false,
            dbColumn: "content_hash"
        },
        data: { type: "json", required: false, dbColumn: "data" },
        metadata: { type: "json", required: false, dbColumn: "metadata" },
        isPublic: { type: "boolean", required: false, dbColumn: "is_public" },
        disabled: { type: "boolean", required: false, dbColumn: "disabled" },
        disabledAt: { type: "date", required: false, dbColumn: "disabled_at" },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" },
        ownerId: { type: "uuid", required: false, dbColumn: "owner_id" },
        organizationId: {
            type: "uuid",
            required: false,
            dbColumn: "organization_id"
        },
        updatedBy: { type: "uuid", required: false, dbColumn: "updated_by" }
    };
}

// Define SourceChunkModel
export class SourceChunkModel extends SupabaseModel {
    static TABLE_NAME = "source_chunk";
    static TABLE_FIELDS = {
        id: { type: "uuid", required: true, dbColumn: "id" },
        sourceId: { type: "uuid", required: false, dbColumn: "source_id" },
        chunkNextId: {
            type: "uuid",
            required: false,
            dbColumn: "chunk_next_id"
        },
        chunkPrevId: {
            type: "uuid",
            required: false,
            dbColumn: "chunk_prev_id"
        },
        data: { type: "json", required: false, dbColumn: "data" },
        metadata: { type: "json", required: false, dbColumn: "metadata" },
        isPublic: { type: "boolean", required: false, dbColumn: "is_public" },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" },
        organizationId: {
            type: "uuid",
            required: false,
            dbColumn: "organization_id"
        }
    };
}

// Define SourceRelationshipModel
export class SourceRelationshipModel extends SupabaseModel {
    static TABLE_NAME = "source_relationship";
    static TABLE_FIELDS = {
        sourceId: {
            type: "uuid",
            required: true,
            dbColumn: "source_id"
        },
        relatedSourceId: {
            type: "uuid",
            required: true,
            dbColumn: "related_source_id"
        },
        relationshipType: {
            type: "string",
            required: false,
            dbColumn: "relationship_type"
        },
        metadata: { type: "json", required: false, dbColumn: "metadata" },
        isPublic: { type: "boolean", required: false, dbColumn: "is_public" },
        disabled: { type: "boolean", required: false, dbColumn: "disabled" },
        disabledAt: { type: "date", required: false, dbColumn: "disabled_at" },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" },
        ownerId: { type: "uuid", required: false, dbColumn: "owner_id" },
        organizationId: {
            type: "uuid",
            required: false,
            dbColumn: "organization_id"
        }
    };

    static async saveToSupabase(
        supabase,
        instance,
        onConflict = ["sourceId", "relatedSourceId"]
    ) {
        return super.saveToSupabase(supabase, instance, onConflict);
    }

    static async upsertToSupabase(
        supabase,
        instances,
        onConflict = ["sourceId", "relatedSourceId"],
        idColumn = "sourceId"
    ) {
        return super.upsertToSupabase(
            supabase,
            instances,
            onConflict,
            idColumn
        );
    }

    static async fetchFromSupabase(
        supabase,
        value = null,
        idColumn = "sourceId"
    ) {
        return super.fetchFromSupabase(supabase, value, idColumn);
    }

    static async existsInSupabase(
        supabase,
        value = null,
        idColumn = "sourceId"
    ) {
        return super.existsInSupabase(supabase, value, idColumn);
    }

    static async deleteFromSupabase(
        supabase,
        value = null,
        idColumn = "sourceId"
    ) {
        return super.deleteFromSupabase(supabase, value, idColumn);
    }
}

File: source/models/supabase/sources.py
from datetime import datetime
from typing import ClassVar, Dict, Optional
from uuid import UUID
from pydantic import Field

# from enum import Enum
from supabase import Client
from supabase.client import AsyncClient

from source.helpers.json_exportable_enum import JSONExportableEnum
from source.models.supabase.supabase_model import SupabaseModel


class SourceTypeEnum(str, JSONExportableEnum):
    DOCUMENT = "document"
    WEBPAGE = "webpage"
    WEBSITE = "website"
    VIDEO = "video"
    AUDIO = "audio"
    IMAGE = "image"
    TOPIC = "topic"
    CONCEPT = "concept"
    COLLECTION = "collection"


class SourceModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "source"
    id: Optional[UUID] = Field(default=None)
    original_source: Optional[str] = Field(default=None)
    resolved_source: Optional[str] = Field(default=None)
    type: Optional[SourceTypeEnum] = Field(default=None)
    title: Optional[str] = Field(default=None)
    lang: Optional[str] = Field(default=None)
    content_hash: Optional[str] = Field(default=None)
    data: Optional[Dict] = Field(default=None)
    metadata: Optional[Dict] = Field(default=None)
    is_public: Optional[bool] = Field(default=False)
    disabled: bool = Field(default=False)
    disabled_at: Optional[datetime] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    owner_id: Optional[UUID] = Field(default=None)
    organization_id: Optional[UUID] = Field(default=None)
    updated_by: Optional[UUID] = Field(default=None)


class SourceChunkModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "source_chunk"
    id: Optional[UUID] = Field(default=None)
    source_id: Optional[UUID] = Field(default=None)
    chunk_next_id: Optional[UUID] = Field(default=None)
    chunk_prev_id: Optional[UUID] = Field(default=None)
    data: Optional[Dict] = Field(default=None)
    metadata: Optional[Dict] = Field(default=None)
    is_public: Optional[bool] = Field(default=False)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    organization_id: Optional[UUID] = Field(default=None)


class SourceRelationshipModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "source_relationship"
    source_id: UUID
    related_source_id: UUID
    relationship_type: Optional[str] = Field(default=None)
    metadata: Optional[Dict] = Field(default=None)
    is_public: Optional[bool] = Field(default=False)
    disabled: bool = Field(default=False)
    disabled_at: Optional[datetime] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    owner_id: Optional[UUID] = Field(default=None)
    organization_id: Optional[UUID] = Field(default=None)

    # @field_validator("metadata", mode="before")
    # def validate_metadata(cls, v):
    #     if isinstance(v, str):
    #         return json.loads(v)
    #     elif isinstance(v, dict):
    #         return json.dumps(v)
    #     return v

    @classmethod
    async def save_to_supabase(
        cls,
        supabase: AsyncClient,
        instance,
        on_conflict=["source_id", "related_source_id"],
    ):
        return await super(SourceRelationshipModel, cls).save_to_supabase(
            supabase, instance, on_conflict
        )

    @classmethod
    async def upsert_to_supabase(
        cls,
        supabase: AsyncClient,
        instances,
        on_conflict=["source_id", "related_source_id"],
        id_column="source_id",
    ):
        return await super(SourceRelationshipModel, cls).upsert_to_supabase(
            supabase, instances, on_conflict, id_column
        )

    @classmethod
    async def fetch_from_supabase(
        cls, supabase: AsyncClient, value=None, id_column="source_id"
    ):
        return await super(SourceRelationshipModel, cls).fetch_from_supabase(
            supabase, value=value, id_column=id_column
        )

    @classmethod
    async def exists_in_supabase(
        cls, supabase: AsyncClient, value=None, id_column="source_id"
    ):
        return await super(SourceRelationshipModel, cls).exists_in_supabase(
            supabase, value=value, id_column=id_column
        )

    @classmethod
    async def delete_from_supabase(
        cls, supabase: AsyncClient, value=None, id_column="source_id"
    ):
        return await super(SourceRelationshipModel, cls).delete_from_supabase(
            supabase, value=value, id_column=id_column
        )

    @classmethod
    def save_to_supabase_sync(
        cls,
        supabase: Client,
        instance,
        on_conflict=["source_id", "related_source_id"],
    ):
        return super(SourceRelationshipModel, cls).save_to_supabase_sync(
            supabase, instance, on_conflict
        )

    @classmethod
    def upsert_to_supabase_sync(
        cls,
        supabase: Client,
        instances,
        on_conflict=["source_id", "related_source_id"],
        id_column="source_id",
    ):
        return super(SourceRelationshipModel, cls).upsert_to_supabase_sync(
            supabase, instances, on_conflict, id_column
        )

    @classmethod
    def fetch_from_supabase_sync(
        cls, supabase: Client, value=None, id_column="source_id"
    ):
        return super(SourceRelationshipModel, cls).fetch_from_supabase_sync(
            supabase, value=value, id_column=id_column
        )

    @classmethod
    def exists_in_supabase_sync(
        cls, supabase: Client, value=None, id_column="source_id"
    ):
        return super(SourceRelationshipModel, cls).exists_in_supabase_sync(
            supabase, value=value, id_column=id_column
        )

    @classmethod
    def delete_from_supabase_sync(
        cls, supabase: Client, value=None, id_column="source_id"
    ):
        return super(SourceRelationshipModel, cls).delete_from_supabase_sync(
            supabase, value=value, id_column=id_column
        )
